app/services/split_timeslices.rb

```diff
@@ -0,0 +1,167 @@
+# frozen_string_literal: true
+
+require_dependency "#{Rails.root}/lib/course_revision_updater"
+require_dependency "#{Rails.root}/lib/timeslice_manager"
+require_dependency "#{Rails.root}/lib/timeslice_cleaner"
+require_dependency "#{Rails.root}/lib/data_cycle/update_debugger"
+require_dependency "#{Rails.root}/lib/revision_scanner"
+
+#= Implements the adaptive timeslice splitting strategy.
+# The algorithm starts with a default timeslice, fetches revisions for that period
+# and, if the timeslice exceeds the threshold, recursively splits it until all
+# timeslices are within limits.
+class SplitTimeslices
+  def initialize(course, update_service: nil)
+    @course = course
+    @timeslice_manager = TimesliceManager.new(@course)
+    @timeslice_cleaner = TimesliceCleaner.new(@course)
+    @revision_updater = CourseRevisionUpdater.new(@course, update_service:)
+    @wikidata_stats_updater = UpdateWikidataStatsTimeslice.new(@course) if wikidata
+  end
+
+  # Max desirable number of revisions per timeslice
+  REVISION_THRESHOLD = 10000
+
+  # Given a wiki and limites of the timeslice, it fetches data for it
+  # (maybe splitting the timeslice if too many revisions) and
+  # returns an array with start dates for timeslices that were processed.
+  # start_date and end_date are the limits of the timeslice records
+  def handle(wiki, start_date, end_date, only_new: true)
+    fetch_only_revisions(wiki, start_date, end_date)
+    if too_many_revisions?(wiki) && splittable?(start_date, end_date)
+      split_timeslice(wiki, start_date, end_date)
+    else
+      log_processing(wiki, start_date, end_date, only_new)
+      # Ensure course wiki timeslice exists for course, wiki and dates
+      @timeslice_manager.maybe_create_course_wiki_timeslice(wiki.id, start_date, end_date)
+      fetch_revisions(wiki, start_date, end_date, only_new:)
+      maybe_fetch_wikidata_stats(wiki)
+      process_timeslices(wiki) if !only_new | new_data?(wiki)
+      [start_date]
+    end
+  end
+
+  private
+
+  def too_many_revisions?(wiki)
+    @revisions[wiki][:revisions].count(&:scoped) > REVISION_THRESHOLD
+  end
+
+  # A timeslice is not split if its duration is an odd number of seconds.
+  # This is because the revisions API expects start and end times in
+  # YYYY-MM-DD HH:MM:SS format, which does not allow fractions of a second.
+  def splittable?(start_date, end_date)
```

Comment:
This odd restriction seems... odd. Are there situations where a default-sized timeslice is not splittable because of this? It seems like we could still split a timeslice with odd seconds, just by putting the odd second into one or the other arbitrarily.